src/gateway/router/routes/registerRoute.ts

```diff
@@ -0,0 +1,108 @@
+import Router from '@koa/router';
+import { evalType } from '../../tasks/contractsMetadata';
+import { BUNDLR_NODE2_URL } from '../../../constants';
+import { DataItem } from 'arbundles';
+import rawBody from 'raw-body';
+import { Tag } from 'arweave/node/lib/transaction';
+import { sleep } from 'warp-contracts';
+import { updateCache } from '../../updateCache';
+
+export async function registerRoute(ctx: Router.RouterContext) {
+  const { logger, gatewayDb, arweave, bundlr } = ctx;
+
+  const rawDataItem: Buffer = await rawBody(ctx.req);
+  const dataItem = new DataItem(rawDataItem);
+
+  try {
+    const isValid = await dataItem.isValid();
+    if (!isValid) {
+      throw new Error(`Data item binary is not valid.`);
+    }
+
+    const areContractTagsValid = await verifyContractTags(dataItem);
+    if (!areContractTagsValid) {
+      throw new Error(`Contract tags are not valid.`);
+    }
+
+    const bundlrResponse = await bundlr.uploader.transactionUploader(dataItem);
+
+    if (
+      bundlrResponse.status !== 200 ||
+      !bundlrResponse.data.public ||
+      !bundlrResponse.data.signature ||
+      !bundlrResponse.data.block
+    ) {
+      throw new Error(
+        `Bundlr did not upload transaction correctly. Bundlr responded with status ${bundlrResponse.status}.`
+      );
+    }
+
+    logger.debug('Data item successfully bundled.', {
+      id: bundlrResponse.data.id,
+    });
+
+    const srcTxId = dataItem.tags.find((d) => d.name == 'Contract-Src')!.value;
+    const initStateRaw = dataItem.tags.find((d) => d.name == 'Init-State')!.value;
+    const initState = JSON.parse(initStateRaw);
+    const type = evalType(initState);
+    const ownerAddress = await arweave.wallets.ownerToAddress(bundlrResponse.data.public);
+    const contentType = dataItem.tags.find((d) => d.name == 'Content-Type')!.value;
+
+    const insert = {
+      contract_id: bundlrResponse.data.id,
+      src_tx_id: srcTxId,
+      init_state: initState,
+      owner: ownerAddress,
+      type: type,
+      pst_ticker: type == 'pst' ? initState?.ticker : null,
+      pst_name: type == 'pst' ? initState?.name : null,
+      block_height: null,
+      block_timestamp: null,
+      content_type: contentType,
+      contract_tx: null,
+      bundler_contract_tx_id: bundlrResponse.data.id,
+      bundler_contract_node: BUNDLR_NODE2_URL,
```

Comment:
It'll be better to extract it from bundlr rater than hardcoding (not critical)

Reply:
Not from the response, but from the bundlr object (should be available under ctx.bundlr, and then bundlr.api (similar to arweave.api))

But this should be probably refactored in all places, not sure if that is a critical "issue" now ;-)

Reply:
ah, ok. let's leave it for now, I don't want this pr to be too dangerous to install :))))